### PR TITLE
Fix linking of OS-provided zlib

### PR DESCRIPTION
**Pull request by directhex**
This is a reduced version of https://github.com/mono/mono/pull/16091 without the Z_PREFIX stuff

### CODE REVIEW

File: mono/profiler/Makefile.am [hunk 1]
```diff
@@ -100,7 +100,7 @@ endif
 if HAVE_STATIC_ZLIB
 zlib_dep = $(STATIC_ZLIB_PATH)
 else
-if HAVE_ZLIB
+if HAVE_SYS_ZLIB
 # The log profiler uses zlib for output compression when available.
 zlib_dep = -lz
 else
```
Comment (akoeplinger) [inline mono/profiler/Makefile.am]: `mono/profiler/log.c` currently checks `HAVE_SYS_ZLIB` so we need to add both the bundled sources here in the Makefile as well as including `"zlib.h"` when we have no system zlib.

Comment (akoeplinger) [reply mono/profiler/Makefile.am]: (and disable it for HOST_WASM in case we build the profiler there)

File: mono/mini/Makefile.am.in [hunk 2]
```diff
@@ -244,7 +244,7 @@ LLVMMONOF=$(LLVM_LIBS) $(LLVM_LDFLAGS)
 endif
 
 if !HOST_WIN32
-if HAVE_ZLIB
+if HAVE_SYS_ZLIB
```
Comment (akoeplinger) [inline mono/mini/Makefile.am.in]: do you know why mini/ has this zlib logic here? It doesn't use any zlib pieces from what I can see. Can you try removing this?

File: mono/profiler/Makefile.am (hunk 1 above)
Comment (directhex) [reply mono/profiler/Makefile.am]: I disagree on this point. All of log.c's lib consumption is gated behind HAVE_SYS_ZLIB. Adding new support for bundled or static zlib to the log profiler is out of scope for this PR.

**Comment by akoeplinger** [comment]
@monojenkins backport 2019-08

**Pull request merged**